File: QytetetJava/src/InterfazTextualQytetet/ControladorQytetet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package InterfazTextualQytetet;

import modeloqytetet.*;
import java.util.ArrayList;
import java.util.Scanner;
import java.util.Map;

/**
 *
 * @author fran
 */
public class ControladorQytetet {

    private static final Scanner IN = new Scanner(System.in);
    private Qytetet juego;
    private Jugador jugador;
    private Casilla casilla;
    private final VistaTextualQytetet vista;

    public ControladorQytetet() {
        vista = new VistaTextualQytetet();
    }

    private void pausa() {
        vista.mostrar("Pulsa una tecla para continuar...");
        IN.nextLine();
    }

    public void inicializacionJuego() {
        juego = Qytetet.getInstance();
        ArrayList<String> nombres = vista.obtenerNombreJugadores();
        juego.inicicializarJuego(nombres);
        jugador = juego.getJugadorActual();
        casilla = juego.getJugadorActual().getCasillaActual();

        vista.mostrar("Tablero: " + juego.getTablero().toString()
                + "\nComienza el jugador: " + jugador.toString());
//                + "\nCasilla actual: " + casilla.toString());
        this.pausa();
    }

    public void desarrolloJuego() {
        while (true) {
            this.estadoActual();
            if (jugador.isEncarcelado()) {
                this.intentarSalirCarcel();
            }
            if (!jugador.isEncarcelado()) {
                vista.mostrar("Tirando dado...");
                espera();
                boolean tienePropietario = juego.jugar();
                casilla = jugador.getCasillaActual();
                vista.mostrar("Valor del dado: " + juego.getValorDado());
                vista.mostrar("Has caido en: " + casilla.toString());
                this.pausa();
                this.estadoActual();

                if (!jugador.isEncarcelado()) {
                    switch (casilla.getTipo()) {
                        case SORPRESA:
                            vista.mostrar(juego.getCartaActual().toString() + "\nAplicando Sopresa...");
                            this.espera();
                            juego.aplicarSorpresa();
                            if (juego.getCartaActual().getTipo() == TipoSorpresa.IRACASILLA) {
                                casilla = jugador.getCasillaActual();
                                vista.mostrar("Nueva " + casilla.toString());
                                if (casilla.getTipo() == TipoCasilla.CARCEL) {
                                    break;
                                }
                            } else {
                                jugador = juego.getJugadorActual();
                                this.estadoActual();
                                break;
                            }//Si la sorpresa lleva al jugador a una calle, entrara en case CALLE
                        case CALLE:
                            if (tienePropietario) {
                                vista.mostrar("\nPagando alquiler...");
                                this.espera();
                            } else {
                                if (vista.elegirQuieroComprar() == 0) {
                                    if (juego.comprarTituloPropiedad()) {
                                        vista.mostrar("\nHas comprado el Titulo de propiedad: "
                                                + casilla.getTitulo().getNombre());
                                    } else {
                                        vista.mostrar("\nNo puedes comprar la propiedad");
                                    }
                                }
                            }
                            this.estadoActual();
                            break;
                    }
                } else {
                    vista.mostrar("Has caido en la casilla del juez,"
                            + "vas directamente a la carcel.");
                    pasaTurno();
                }
            }
            if (!jugador.isEncarcelado()){
            if (jugador.tengoPropiedades()){
                gestionInmobiliaria();
            } else {
                vista.mostrar("No tienes propiedades");
            }
            }else{
                vista.mostrar("Estas en la carcel");
            }
//            this.estadoActual();
            pasaTurno();
        }
    }

    private void gestionInmobiliaria() {
        int opcion;
        do {
            opcion = vista.menuGestionInmobiliaria();
            if (jugador.tengoPropiedades() || opcion == 0) {
                ArrayList<Casilla> noHipotecadas = juego.propiedadesHipotecadasJugador(false);
                ArrayList<Casilla> Hipotecadas = juego.propiedadesHipotecadasJugador(true);
                Casilla modCasilla;
                switch (opcion) {
                    case 0:
                        break;
                    case 1:
                        if (!noHipotecadas.isEmpty()) {
                            modCasilla = elegirPropiedad(noHipotecadas);
                            if (!(modCasilla == null)) {
                                if (juego.edificarCasa(modCasilla)) {
                                    vista.mostrar("Has edificado una casa en "
                                            + modCasilla.getTitulo().getNombre());
                                } else {
                                    vista.mostrar("No has podido edificar la casa.");
                                }
                            }
                        } else {
                            vista.mostrar("Todas tus propiedades estan hipotecadas");
                            pausa();
                        }
                        break;
                    case 2:
                        if (!noHipotecadas.isEmpty()) {
                            modCasilla = elegirPropiedad(noHipotecadas);
                            if (!(modCasilla == null)) {
                                if (juego.edificarHotel(modCasilla)) {
                                    vista.mostrar("Has edificado un hotel en "
                                            + modCasilla.getTitulo().getNombre());
                                } else {
                                    vista.mostrar("No has podido edificar el hotel.");
                                }
                            }
                        } else {
                            vista.mostrar("Todas tus propiedades estan hipotecadas");
                            pausa();
                        }
                        break;
                    case 3:
                        if (!noHipotecadas.isEmpty()) {
                            modCasilla = elegirPropiedad(noHipotecadas);
                            if (!(modCasilla == null)) {
                                if (juego.venderPropiedad(modCasilla)) {
                                    vista.mostrar("Has vendido la propiedad "
                                            + modCasilla.getTitulo().getNombre());
                                } else {
                                    vista.mostrar("No has podido vender la propiedad");
                                }
                            }
                        } else {
                            vista.mostrar("Todas tus propiedades estan hipotecadas");
                            pausa();
                        }
                        break;
                    case 4:
                        if (!noHipotecadas.isEmpty()) {
                            modCasilla = elegirPropiedad(noHipotecadas);
                            if (!(modCasilla == null)) {
                                if (juego.hipotecarPropiedad(modCasilla)) {
                                    vista.mostrar("Has hipotecado la propiedad "
                                            + modCasilla.getTitulo().getNombre());
                                } else {
                                    vista.mostrar("No has podido hipotecar la propiedad");
                                }
                            }
                        } else {
                            vista.mostrar("Todas tus propiedades estan hipotecadas");
                            pausa();
                        }
                        break;
                    case 5:
                        if (!Hipotecadas.isEmpty()) {
                            modCasilla = elegirPropiedad(Hipotecadas);
                            if (!(modCasilla == null)) {
                                if (juego.cancelarHipoteca(modCasilla)) {
                                    vista.mostrar("Has cancelado la hipoteca de "
                                            + modCasilla.getTitulo().getNombre());
                                } else {
                                    vista.mostrar("No has podido cancelar la hipoteca");
                                }
                            }
                        } else {
                            vista.mostrar("No tienes propiedades hipotecadas");
                            pausa();
                        }
                        break;
                    default:
                        break;
                }
                estadoActual();
            } else {
                vista.mostrar("No te quedan propiedades");
                pausa();
            }
        } while (opcion != 0);
    }

    private void pasaTurno() {
        vista.mostrar("Turno siguiente jugador...");
        juego.siguienteJugador();
        jugador = juego.getJugadorActual();
        casilla = jugador.getCasillaActual();
        this.espera();
    }

    private Boolean bancarrota() {
        return jugador.getSaldo() < 0;
    }

    private void gameOver() {
        int pos = 1;
        vista.mostrar("Game Over! Jugador " + jugador.getNombre() + " en bancarrota");
        vista.mostrar("Ranking segun el capital: ");
        Map<String, Integer> map = juego.obtenerRanking();
        for (Map.Entry<String, Integer> entry : map.entrySet()) {
            vista.mostrar(pos + ". " + entry.getKey() + " -- " + entry.getValue());
            pos++;
        }
        System.exit(0);
    }

    private void estadoActual() {
        vista.mostrar("**-JUGADOR-**"
                + "\nNombre: " + jugador
                + "\nCasilla actual: " + casilla.getNumeroCasilla()
                + "\nEncarcelado: " + jugador.isEncarcelado()
                + "\nCarta libertad: " + jugador.tengoCartaLibertad()
                + "\nSaldo actual: " + jugador.getSaldo()
        );
        if (this.bancarrota()) {
            this.gameOver();
        }
        this.pausa();
    }

    private void espera() {
        try {
            Thread.sleep(1000);
        } catch (InterruptedException e) {
            System.out.println("Continuando...");
        }
    }

    private void intentarSalirCarcel() {
        boolean libre = false;
        int eleccion = vista.menuSalirCarcel();

        if (eleccion == 0) {
            libre = juego.intentarSalirCarcel(MetodoSalirCarcel.TIRANDODADO);
            vista.mostrar("Valor dado: " + juego.getValorDado());
        } else {
            libre = juego.intentarSalirCarcel(MetodoSalirCarcel.PAGANDOLIBERTAD);
        }
        if (libre) {
            vista.mostrar("Hurra!! Has salido de la carcel");
        } else {
            vista.mostrar("Lo siento. Sigues en la carcel");
        }
        this.pausa();
    }

    public Casilla elegirPropiedad(ArrayList<Casilla> propiedades) {
        //este metodo toma una lista de propiedades y genera una lista de strings, con el numero y nombre de las propiedades
        //luego llama a la vista para que el usuario pueda elegir.
        vista.mostrar("\tCasilla\tTitulo");
        int seleccion;
        ArrayList<String> listaPropiedades = new ArrayList();
        for (Casilla casilla : propiedades) {
            listaPropiedades.add("\t" + casilla.getNumeroCasilla() + "\t" + casilla.getTitulo().getNombre());
        }
        seleccion = vista.menuElegirPropiedad(listaPropiedades);
        if (seleccion == listaPropiedades.size()) {
            return null;
        } else {
            return propiedades.get(seleccion);
        }
    }

}

File: QytetetRuby/lib/controlador_qytetet.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module InterfazTextualQytetet
  require_relative 'qytetet'
  require_relative 'vista_textual_qytetet.rb'
  
  class ControladorQytetet
    def initialize
      @juego = nil
      @jugador = nil
      @casilla = nil
      @vista   = VistaTextualQytetet.new
    end
    
    def inicializacion_juego
      @juego = ModeloQytetet::Qytetet.instance
      @juego.inicializar_juego(@vista.obtener_nombre_jugadores)
      @jugador = @juego.jugador_actual
      @casilla = @juego.jugador_actual.casilla_actual
      #EXAMEN-INICIO
      @jugador.carta_loteria = @juego.mazo.pop
      @jugador.propiedades << @juego.tablero.obtener_casilla_numero(1).titulo_propiedad
      @jugador.propiedades << @juego.tablero.obtener_casilla_numero(2).titulo_propiedad
      @jugador.propiedades << @juego.tablero.obtener_casilla_numero(4).titulo_propiedad
      @vista.mostrar("Carta loteria: #{@jugador.carta_loteria}")
      #EXAMEN-FIN
      @vista.mostrar("Tablero: #{@juego.tablero}" \
          "\nComienza el jugador: #{@jugador}"\
          + "\nCasilla actual: #{@casilla}")
      pausa
    end
    
    def estado_actual
      @vista.mostrar("**-JUGADOR-**"\
          "\nNombre: #{@jugador}" \
          "\nCasilla actual: #{@casilla.numero_casilla}" \
          "\nEncarcelado: #{@jugador.encarcelado}" \
          "\nCarta libertad: #{@jugador.tengo_carta_libertad}"\
          "\nCarta loteria: #{@jugador.tengo_carta_loteria}"\
          "\nSaldo actual: #{@jugador.saldo}" 
      );
      if bancarrota 
        @vista.mostrar("Jugador #{@jugador.nombre} en bancarrota.")
        game_over
      end
      pausa
    end
      
    def bancarrota
      @jugador.saldo < 0
    end
      
    def game_over
      @vista.mostrar("GAME OVER!!!.\nRanking:\n#{@juego.obtener_ranking}")
    end
      
    def desarrollo_juego
      loop do
        estado_actual
        if !intentar_salir_carcel
          @vista.mostrar('Sigues en la carcel, pasas turno.')
          pausa
        else
          @vista.mostrar("Tirando dado...")
          espera
          estado = @juego.jugar
          @casilla = @jugador.casilla_actual
          @vista.mostrar("\nValor del dado: #{estado[:dado]}")
          @vista.mostrar("Has caido en la casilla numero: #{@casilla.numero_casilla}"\
              "\nTipo: #{@casilla.tipo}")
          pausa
          estado_actual
          
          if !@jugador.encarcelado
            case @casilla.tipo
            when ModeloQytetet::TipoCasilla::SORPRESA
              @vista.mostrar("#{@juego.carta_actual}\nAplicando Sorpresa...")
              @juego.aplicar_sorpresa
              @jugador = @juego.jugador_actual #Si el jugador se ha convertido en especulador
              estado_actual
              if @juego.carta_actual == ModeloQytetet::TipoSorpresa::IRACASILLA
                @casilla = @jugador.casilla_actual
                @vista.mostrar "Nueva casilla: #{@casilla}"
              end
            when ModeloQytetet::TipoCasilla::CALLE
              @vista.mostrar("Has caido en la calle:\n#{@casilla}")
              if estado[:tiene_propietario]
                @vista.mostrar("\nPagando alquiler...")
                espera
              else
                elegido = @vista.elegir_comprar_propiedad 
                comprado = @juego.comprar_titulo_propiedad if elegido
                if comprado
                  @vista.mostrar("Has comprado el Titulo de propiedad: "\
                      "#{@casilla.titulo_propiedad.nombre}") 
                else
                  @vista.mostrar("\nNo puedes comprar la propiedad")
                end
              end
              estado_actual
            end
          else
            @vista.mostrar("Has caido en la casilla del juez,"\
                "vas directamente a la carcel.")
          end
        end
        if @jugador.tengo_propiedades && !@jugador.encarcelado
          gestion_inmobiliaria
        else
          @vista.mostrar("No tienes propiedades")
        end
        pasa_turno
      end
    end
    
    def pasa_turno
      @vista.mostrar("\n*****PASA DE TURNO*****\n")
      @vista.mostrar("\nTurno siguiente jugador...")
      @juego.siguiente_jugador
      @jugador = @juego.jugador_actual
      @casilla = @jugador.casilla_actual
      espera
    end
    
    def gestion_inmobiliaria
      begin
        opcion = @vista.menu_gestion_inmobiliaria
        if @jugador.tengo_propiedades || opcion == 0
          propiedades = @juego.propiedades_hipotecadas_jugador(opcion == 5)
          propiedades.each{|p| puts p}
          pausa
          prop = elegir_propiedad(propiedades) unless opcion == 0 || opcion == 6
          case opcion
#          when 0 then break
          when 1 then @juego.edificar_casa(prop)
          when 2 then @juego.edificar_hotel(prop)
          when 3 then @juego.vender_propiedad(prop)
          when 4 then @juego.hipotecar_propiedad(prop)
          when 5 then @juego.cancelar_hipoteca(prop)
          when 6 then @juego.irme_a_vivir_a_cancun
          end
          @vista.mostrar(prop)
        else
          @vista.mostrar('No te quedan propiedades')
        end
        
      end while opcion != 0
    end
    def intentar_salir_carcel
      return true if !@jugador.encarcelado
      metodo = case @vista.menu_salir_carcel
      when 0 then MetodoSalirCarcel::TIRANDODADO
      when 1 then MetodoSalirCarcel::PAGANDOLIBERTAD
      end
      @juego.intentar_salir_carcel(metodo)
    end
    
    def espera
      sleep(1)
    end
    
    def pausa
      puts 'Pulsa enter para continuar: '
      gets
    end
      
    def elegir_propiedad(propiedades) # lista de propiedades a elegir
      @vista.mostrar("\tCasilla\tTitulo")
      lista = propiedades.map { |p| "#{p.numero_casilla}  \t#{p.titulo_propiedad.nombre}" }
      
      propiedades.at(@vista.menu_elegir_propiedad(lista))
    end
    
    def main
      inicializacion_juego
      desarrollo_juego
      fin_juego
    end
  end
end
